File: BinNode/BinNode.c
// BinNode.c
#include "BinNode.h"

#include <stdio.h>
#include <stdlib.h>

//method to create pointer to binnode struct and initialize its fields
BinNode* CreateBinNode(void* value) {
    //allocate memory
    BinNode* node = (BinNode*)malloc(sizeof(BinNode));

    //initialize its fields
    node->value = value;
    node->left = NULL;
    node->right = NULL;

    //return the pointer
    return node;
}

//set value
void SetBinNodeValue(BinNode* node, void* value) { node->value = value; }

//get value
void* GetBinNodeValue(const BinNode* node) { return node->value; }

//get left son
BinNode* GetLeft(const BinNode* node) { return node->left; }

//get right son
BinNode* GetRight(const BinNode* node) { return node->right; }

//set left son
void SetLeft(BinNode* node, BinNode* left) { node->left = left; }

//set right son
void SetRight(BinNode* node, BinNode* right) { node->right = right; }

//is parent has left son
int HasLeft(const BinNode* node) { return node->left != NULL; }

//is parent has right son
int HasRight(const BinNode* node) { return node->right != NULL; }

//is binnode is leaf
int IsLeaf(const BinNode* node) { return node->left == NULL && node->right == NULL; }

//print binnode
void printIntBin(const BinNode* node) { if (node != NULL) { printf("%d ", *(int*)node->value); } }

File: src/main/java/org/example/personalpractice/BinNode.java
package org.example.personalpractice;

public class BinNode<T> {
    private BinNode<T> left;  // חיבור לילד השמאלי
    private T value;          // הערך של הצומת
    private BinNode<T> right; // חיבור לילד הימני

    public BinNode(BinNode<T> left, T value, BinNode<T> right) {
        this.left = left;
        this.value = value;
        this.right = right;
    }

    public BinNode(T x) {
        this.left=null;
        this.value=x;
        this.right=null;
    }

    public BinNode<T> getLeft() {
        return left;
    }

    public void setLeft(BinNode<T> left) {
        this.left = left;
    }

    public T getValue() {
        return value;
    }

    public BinNode<T> getRight() {
        return right;
    }

    public void setRight(BinNode<T> right) {
        this.right = right;
    }


    public void setValue(T value) {
        this.value = value;
    }

    // פעולה בוליאנית שבודקת אם יש חוליה ימנית
    public boolean hasRight() {
        return this.right != null;
    }

    // פעולה בוליאנית שבודקת אם יש חוליה שמאלית
    public boolean hasLeft() {
        return this.left != null;
    }

    // פעולה בוליאנית שבודקת אם החוליה היא עלה (אין שמאלי ואין ימני)
    public boolean isLeaf() {
        return this.left == null && this.right == null;
    }
}
